### ISSUE TITLE
api doc and c# 7 value tuples

### ISSUE THREAD

**Issue by agnauck** [statiqdev/Statiq#749]
API doc generation is failing on generics with c# 7 value tuples.
Its throwing a System.ArgumentOutOfRangeException

This interface can be used to reproduce the exception

```
using System;
using System.Threading.Tasks;

public interface IFoo
{
   Task<(string, string)> Bar();
}
```

**Comment by daveaglick** [comment]
It looks like this might be resolved in 2.0 (see #750) - can you give that a try when you get the chance and report back?

**Comment by agnauck** [comment]
Just ran a test with 2.0.0. Still getting the same exception,

```
Exception during execution: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Count cannot be less than zero.
Parameter name: count)) (Count cannot be less than zero.
Parameter name: count) ---> System.ArgumentOutOfRangeException: Count cannot be less than zero.
Parameter name: count
   at System.String.Remove(Int32 startIndex, Int32 count)
   at Wyam.Docs.Extensions.GetTypeLink(IExecutionContext context, IMetadata metadata, String name, Boolean linkTypeArguments)
   at AspNetCore._Shared_Section__Methods.ExecuteAsync() in C:/Users/Alex/.nuget/packages/Wyam.Docs.Samson.2.0.0/content/Shared/Section/_Methods.cshtml:line 26
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.RenderPartialCoreAsync(String partialViewName, Object model, ViewDataDictionary viewData, TextWriter writer)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.PartialAsync(String partialViewName, Object model, ViewDataDictionary viewData)
   at Microsoft.AspNetCore.Mvc.Rendering.HtmlHelperPartialExtensions.Partial(IHtmlHelper htmlHelper, String partialViewName, Object model, ViewDataDictionary viewData)
   at AspNetCore._Shared_Kind__NamedType.ExecuteAsync() in C:/Users/Alex/.nuget/packages/Wyam.Docs.Samson.2.0.0/content/Shared/Kind/_NamedType.cshtml:line 155
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.RenderPartialCoreAsync(String partialViewName, Object model, ViewDataDictionary viewData, TextWriter writer)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.PartialAsync(String partialViewName, Object model, ViewDataDictionary viewData)
   at Microsoft.AspNetCore.Mvc.Rendering.HtmlHelperPartialExtensions.Partial(IHtmlHelper htmlHelper, String partialViewName, Object model, ViewDataDictionary viewData)
   at AspNetCore.__ApiLayout.ExecuteAsync() in C:/Users/Alex/.nuget/packages/Wyam.Docs.Samson.2.0.0/content/_ApiLayout.cshtml:line 40
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderLayoutAsync(ViewContext context, ViewBufferTextWriter bodyWriter)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Wyam.Razor.RazorCompiler.RenderPage(RenderRequest request)
   at Wyam.Razor.RazorService.Render(RenderRequest request)
   at Wyam.Razor.Razor.<>c__DisplayClass16_0.<Execute>b__1(IDocument input)
   at Wyam.Common.Execution.TraceExceptionsExtensions.TraceExceptions[TResult](IExecutionContext context, IDocument document, Func`2 func) in C:\Code\Wyam\src\core\Wyam.Common\Execution\TraceExceptionsExtensions.cs:line 50
   at System.Linq.Parallel.SelectQueryOperator`2.SelectQueryOperatorResults.GetElement(Int32 index)
   at System.Linq.Parallel.QueryResults`1.get_Item(Int32 index)
   at System.Linq.Parallel.PartitionedDataSource`1.ListContiguousIndexRangeEnumerator.MoveNext(T& currentElement, Int32& currentKey)
   at System.Linq.Parallel.PipelineSpoolingTask`2.SpoolingWork()
   at System.Linq.Parallel.SpoolingTaskBase.Work()
   at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
   at System.Linq.Parallel.QueryTask.<>c.<.cctor>b__10_0(Object o)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   --- End of inner exception stack trace ---
   at System.Linq.Parallel.QueryTaskGroupState.QueryEnd(Boolean userInitiatedDispose)
   at System.Linq.Parallel.AsynchronousChannelMergeEnumerator`1.MoveNextSlowPath()
   at System.Linq.Parallel.AsynchronousChannelMergeEnumerator`1.MoveNext()
   at System.Linq.Parallel.QueryOpeningEnumerator`1.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToArray()
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at System.Collections.Immutable.ImmutableArray.CreateRange[T](IEnumerable`1 items)
   at System.Collections.Immutable.ImmutableArray.ToImmutableArray[TSource](IEnumerable`1 items)
   at Wyam.Core.Execution.ExecutionPipeline.Execute(ExecutionContext context, IEnumerable`1 modules, ImmutableArray`1 inputDocuments) in C:\Code\Wyam\src\core\Wyam.Core\Execution\ExecutionPipeline.cs:line 144
   at Wyam.Core.Execution.ExecutionContext.Execute(IEnumerable`1 modules, IEnumerable`1 inputs, IEnumerable`1 items) in C:\Code\Wyam\src\core\Wyam.Core\Execution\ExecutionContext.cs:line 189
   at Wyam.Core.Modules.Control.If.Execute(IReadOnlyList`1 inputs, IExecutionContext context) in C:\Code\Wyam\src\core\Wyam.Core\Modules\Control\If.cs:line 141
   at Wyam.Core.Execution.ExecutionPipeline.Execute(ExecutionContext context, IEnumerable`1 modules, ImmutableArray`1 inputDocuments) in C:\Code\Wyam\src\core\Wyam.Core\Execution\ExecutionPipeline.cs:line 144
   at Wyam.Core.Execution.ExecutionPipeline.Execute(Engine engine, Guid executionId) in C:\Code\Wyam\src\core\Wyam.Core\Execution\ExecutionPipeline.cs:line 85
   at Wyam.Core.Execution.Engine.Execute() in C:\Code\Wyam\src\core\Wyam.Core\Execution\Engine.cs:line 247
---> (Inner Exception #0) System.ArgumentOutOfRangeException: Count cannot be less than zero.
Parameter name: count
   at System.String.Remove(Int32 startIndex, Int32 count)
   at Wyam.Docs.Extensions.GetTypeLink(IExecutionContext context, IMetadata metadata, String name, Boolean linkTypeArguments)
   at AspNetCore._Shared_Section__Methods.ExecuteAsync() in C:/Users/Alex/.nuget/packages/Wyam.Docs.Samson.2.0.0/content/Shared/Section/_Methods.cshtml:line 26
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.RenderPartialCoreAsync(String partialViewName, Object model, ViewDataDictionary viewData, TextWriter writer)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.PartialAsync(String partialViewName, Object model, ViewDataDictionary viewData)
   at Microsoft.AspNetCore.Mvc.Rendering.HtmlHelperPartialExtensions.Partial(IHtmlHelper htmlHelper, String partialViewName, Object model, ViewDataDictionary viewData)
   at AspNetCore._Shared_Kind__NamedType.ExecuteAsync() in C:/Users/Alex/.nuget/packages/Wyam.Docs.Samson.2.0.0/content/Shared/Kind/_NamedType.cshtml:line 155
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.RenderPartialCoreAsync(String partialViewName, Object model, ViewDataDictionary viewData, TextWriter writer)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.PartialAsync(String partialViewName, Object model, ViewDataDictionary viewData)
   at Microsoft.AspNetCore.Mvc.Rendering.HtmlHelperPartialExtensions.Partial(IHtmlHelper htmlHelper, String partialViewName, Object model, ViewDataDictionary viewData)
   at AspNetCore.__ApiLayout.ExecuteAsync() in C:/Users/Alex/.nuget/packages/Wyam.Docs.Samson.2.0.0/content/_ApiLayout.cshtml:line 40
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderLayoutAsync(ViewContext context, ViewBufferTextWriter bodyWriter)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Wyam.Razor.RazorCompiler.RenderPage(RenderRequest request)
   at Wyam.Razor.RazorService.Render(RenderRequest request)
   at Wyam.Razor.Razor.<>c__DisplayClass16_0.<Execute>b__1(IDocument input)
   at Wyam.Common.Execution.TraceExceptionsExtensions.TraceExceptions[TResult](IExecutionContext context, IDocument document, Func`2 func) in C:\Code\Wyam\src\core\Wyam.Common\Execution\TraceExceptionsExtensions.cs:line 50
   at System.Linq.Parallel.SelectQueryOperator`2.SelectQueryOperatorResults.GetElement(Int32 index)
   at System.Linq.Parallel.QueryResults`1.get_Item(Int32 index)
   at System.Linq.Parallel.PartitionedDataSource`1.ListContiguousIndexRangeEnumerator.MoveNext(T& currentElement, Int32& currentKey)
   at System.Linq.Parallel.PipelineSpoolingTask`2.SpoolingWork()
   at System.Linq.Parallel.SpoolingTaskBase.Work()
   at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
   at System.Linq.Parallel.QueryTask.<>c.<.cctor>b__10_0(Object o)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)<---
```

**Comment by daveaglick** [comment]
I can replicate, looking at this now

**Comment by daveaglick** [comment]
Thanks for the bug report - fix for this will go out with 2.1.0
